### ISSUE TITLE
libglew issues on Windows builds with mingw / cmake

### ISSUE THREAD

**Issue by imaclmaca** [fltk/fltk#226]
During the 1.3.6 rc testing, I attempted to enable GLEW support in the cmake build.
NOTE: I believe that our cmake configuration uses pkg-config to determine libglew support, and my mingw testing environments *do not* have a working pkg-config, so I *expected* this to fail at configure time.
However, in fact this did not fail until link time, when the 2 OGL3 samples in the examples folder failed.

First off, the build failed because, when linking the GLEW examples, it could not locate libGLEW.a. The GLEW lib on mingw* is called libglew32.a (for both 32 or 64-bit builds) however, so it looks like we have asked for "the wrong" lib name at this point. 
I made a copy of libglew32.a to libGLEW.a and the build then found and linked the lib in OK. 
However, the link then failed for want of libopengl32.a in the link command. 
For the "normal" OpenGL examples, we do link "-lopengl32" correctly so this behaviour is peculiar to the GLEW targets.
I tweaked the libs to ensure "-lopengl32" was incorporated and thereafter the two GLEW examples built and ran correctly.

I assume these issues arise because I do not have a working pkg-config in my mingw environments, but thought I'd record the issue here just as a record.

Effect observed on both fltk-1.3 and fltk-1.4 branches.
Builds performed on Win10, using mingw32 / Msys for 32-bit builds and TDM-gcc / Msys for 64-bit builds. Results were "the same" on both build environments.
(With the build "adjusted" manually, the samples build and run correctly.)

**Comment by Albrecht-S** [comment]
Thanks, I'll look into this later...

**Comment by Albrecht-S** [comment]
This issue should be fixed now in the master branch.

I could test the build using my MSYS2 (64-bit) build, it found the libs and could build the executables. However, in my VM I could not run the programs:
```
$ bin/examples/OpenGL3-glut-test.exe
Status: Using GLEW 2.2.0
OpenGL version 1.1.0 supported

This platform does not support OpenGL V3
```
I assume that this is a runtime issue (either a VM problem or I'm missing something else) but the libs are found and linked correctly.

@imaclmaca Can you please try again and report if it works for you now?

**Comment by imaclmaca** [comment]
Hmm, OK.
Re-tried that, in new "build" folders, and both the 32-bit and 64-bit builds did what they did before, that is they built OK, up until the link stage - then the final link of the GLEW targets was missing the OpenGL lib, so failed with...

Scanning dependencies of target OpenGL3-glut-test
[ 98%] Building CXX object examples/CMakeFiles/OpenGL3-glut-test.dir/OpenGL3-glut-test.cxx.obj
[100%] Linking CXX executable ../bin/examples/OpenGL3-glut-test.exe
/MINGW/lib/libGLEW.a(glew.o):glew.c:(.text+0x144ff): undefined reference to `wglGetCurrentDC@0'
/MINGW/lib/libGLEW.a(glew.o):glew.c:(.text+0x1459e): undefined reference to `wglGetCurrentDC@0'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [bin/examples/OpenGL3-glut-test.exe] Error 1
make[1]: *** [examples/CMakeFiles/OpenGL3-glut-test.dir/all] Error 2
make: *** [all] Error 2
 
Manually adding the OpenGL lib to the linker line sorts that, but it would be better if the fltk cmake did.
Note that other GL targets build OK, so this seems to be peculiar to the GLEW targets here.

And I would not discount the possibility that the issue is with the configuration of my systems, rather than any more generic issue with the cake files!

**Comment by Albrecht-S** [comment]
OK, I see that libglew32 is now found, is this correct?

To debug the linker error, can you please post the linker commandline before and after your changes?

Besides that, can you please apply the following patch so we can debug some variables? It would be fortunate if you could post the entire cmake run (the configure/generate part, before the build starts). TIA

```diff
diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
index c358a382c..a5ea61660 100644
--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -155,6 +155,10 @@ else ()
   set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES} ${LIB_GLEW})
 endif ()
 
+fl_debug_var (OPENGL_LIBRARIES)
+fl_debug_var (LIB_GLEW)
+fl_debug_var (REQUIRED_LIBS)
+
 if (OPENGL_FOUND AND LIB_GLEW)
   foreach (src ${OPENGL_SOURCES})
     CREATE_EXAMPLE (${src} ${src}.cxx "${REQUIRED_LIBS}")
```

**Comment by imaclmaca** [comment]
On Mon, 7 Mar 2022 at 21:35, Albrecht Schlosser wrote:

> OK, I see that libglew32 is now found, is this correct?
>

Ah! Good point.
No, in my case, it actually finds libGLEW.a, but that is only because, at
some point in the past, I manually copied libglew32.a to libGLEW.a,
probably as part of working around this issue!

To debug the linker error, can you please post the linker commandline
> before and after your changes?
>
Not quite sure what to post here.
My change is nothing sophisticated, all I am doing is adding "-lopengl32"
to the LDFLAGS in the cmake config (which is taking a hammer to it so that
*all* links get offered OpenGL whether they need it or not!) and of course
that makes the build work. But it is no way a proper fix to this issue.
Without that bodge, the GLEW targets fail for want of ***@***.***
basically.

Besides that, can you please apply the following patch so we can debug some
> variables? It would be fortunate if you could post the entire cmake run
> (the configure/generate part, before the build starts). TIA
>
> diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
> index c358a382c..a5ea61660 100644--- a/examples/CMakeLists.txt+++ b/examples/CMakeLists.txt@@ -155,6 +155,10 @@ else ()
>    set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES} ${LIB_GLEW})
>  endif ()
>  +fl_debug_var (OPENGL_LIBRARIES)+fl_debug_var (LIB_GLEW)+fl_debug_var (REQUIRED_LIBS)+
>  if (OPENGL_FOUND AND LIB_GLEW)
>    foreach (src ${OPENGL_SOURCES})
>      CREATE_EXAMPLE (${src} ${src}.cxx "${REQUIRED_LIBS}")
>
>
Log of the config attached as a zip (I hope!), see below.

**Comment by imaclmaca** [comment]
On Tue, 8 Mar 2022 at 12:55, imm wrote:
>
> On Mon, 7 Mar 2022 at 21:35, Albrecht Schlosser wrote:
>>
>> OK, I see that libglew32 is now found, is this correct?
>
>
> Ah! Good point.
> No, in my case, it actually finds libGLEW.a, but that is only because, at some point in the past, I manually copied libglew32.a to libGLEW.a, probably as part of working around this issue!
>

OK, so I moved aside my renamed libGLEW.a and did a fresh, new build
folder - this did correctly find libglew32.a this time.
However, the GLEW examples still failed to link for want of libopengl32...

**Comment by Albrecht-S** [comment]
> OK, so I moved aside my renamed libGLEW.a and did a fresh, new build
folder - this did correctly find libglew32.a this time.

Good so far, thanks for confirmation.

> Log of the config attached as a zip (I hope!), see below.

Unfortunately there's no zip file.

Regarding the linker command line:

> Not quite sure what to post here.
> My change is nothing sophisticated, all I am doing is adding "-lopengl32" to the LDFLAGS in the cmake config
> However, the GLEW examples still failed to link for want of libopengl32...

OK, I understand. I thought you had found and edited the real linker command and could post it here. I assume you're working with `make` after you created the build files with `cmake`. Then you can do these two steps:
1. run the build until it stops, preferred with '-j1' or equivalent
2. run the build again in VERBOSE mode (should display the command line).
```
$ make OpenGL3test
$ make VERBOSE=ON OpenGL3test
```

Please post only the linker commandline here.

[ Testing ... ] Hmm, I tried it (looks fine under `Linux` but pretty obfuscated under `MSYS2/MinGW`). Anyway, what I found out (YMMV) is that the linker command is split in several files (Makefile, something.txt, something.rsp) which is more than confusing.

This is the relevant part I found (cwd is the build root):
```
$ cat examples/CMakeFiles/OpenGL3test.dir/linklibs.rsp
 ../lib/libfltk_gl.a ../lib/libfltk.a -lopengl32 -lglu32 -lglew32 -lgdiplus -lcomctl32 -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32
```
As you can see the link order is `-lopengl32 -lglu32 -lglew32 ...` which may be causing your issues. It works for me though, maybe because it links the shared libs (here) rather than the static libs (in your case?).

That said, below is another simple patch that can help by changing the link order. After the patch I got:
```
$ cat examples/CMakeFiles/OpenGL3test.dir/linklibs.rsp
 ../lib/libfltk_gl.a ../lib/libfltk.a -lglew32 -lopengl32 -lglu32 -lgdiplus -lcomctl32 -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32
```
with `-lglew32 -lopengl32 -lglu32` which **might** help because `glew32` is now before `opengl32`.

The patch is simple - move `${LIB_GLEW}` before `${OPENGL_LIBRARIES}`:
```
diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
index c358a382c..fc92de28b 100644
--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -152,7 +152,7 @@ if (APPLE AND (NOT OPTION_APPLE_X11))
   endif ()
   set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES})
 else ()
-  set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES} ${LIB_GLEW})
+  set (REQUIRED_LIBS fltk_gl fltk ${LIB_GLEW} ${OPENGL_LIBRARIES})
 endif ()
 
 if (OPENGL_FOUND AND LIB_GLEW)
```

Note that I didn't commit/push this yet because I wanted to wait for your feedback. If you could apply the previous patch above that adds a few `fl_debug_var()` statements as well this could help me to understand what's going on.

Hopefully this fixes your build. Can you confirm?

**Comment by Albrecht-S** [comment]
@imaclmaca Ian, any news on this build issue?

**Comment by Albrecht-S** [comment]
@imaclmaca wrote in a private mail:
> "reordering the glew lib before the GL libs works perfectly. That seems to be the solution."

This has been committed in 345e04c08353407a805e67abbd949995e95e03d8, hence this issue is now fixed.

**Comment by imaclmaca** [comment]
On Wed, 9 Mar 2022, 21:11 Albrecht Schlosser wrote:

> @imaclmaca <https://github.com/imaclmaca> Ian, any news on this build
> issue?
>

I replied earlier (checking: 15:37 apparently) did you not receive it?

Short answer: reordering the glew lib before the GL libs works perfectly.
That seems to be the solution.

--
Ian
From my Fairphone FP3

**Comment by imaclmaca** [comment]
On Tue, 8 Mar 2022 at 16:56, Albrecht Schlosser wrote:

> Log of the config attached as a zip (I hope!), see below.
>
> Unfortunately there's no zip file.
>

Huh! OK, that's a pity. I wonder of the github mail routing strips 'rm off,
as it looks (from here) as if I did send it?
And.. hmm, I can see it on my phone, too.
Oh well, computers are weird...


> That said, below is another simple patch that can help by changing the
> link order. After the patch I got:
>
> $ cat examples/CMakeFiles/OpenGL3test.dir/linklibs.rsp
>  ../lib/libfltk_gl.a ../lib/libfltk.a -lglew32 -lopengl32 -lglu32 -lgdiplus -lcomctl32 -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32
>
> with -lglew32 -lopengl32 -lglu32 which *might* help because glew32 is now
> before opengl32.
>

Yup, that's do it; the gcc static linker is order sensitive by default
(though IIRC there is an option to enable rescanning, but folks are wary of
it... slow? Dunno... May just be that it makes it harder to control the
link order I guess.)

> The patch is simple - move ${LIB_GLEW} before ${OPENGL_LIBRARIES}:
>
> diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
> index c358a382c..fc92de28b 100644
> --- a/examples/CMakeLists.txt
> +++ b/examples/CMakeLists.txt
> @@ -152,7 +152,7 @@ if (APPLE AND (NOT OPTION_APPLE_X11))
>    endif ()
>    set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES})
>  else ()
> -  set (REQUIRED_LIBS fltk_gl fltk ${OPENGL_LIBRARIES} ${LIB_GLEW})
> +  set (REQUIRED_LIBS fltk_gl fltk ${LIB_GLEW} ${OPENGL_LIBRARIES})
>  endif ()
>
>  if (OPENGL_FOUND AND LIB_GLEW)
>
> Note that I didn't commit/push this yet because I wanted to wait for your
> feedback. If you could apply the previous patch above that adds a few
> fl_debug_var() statements as well this could help me to understand what's
> going on.
>
> Hopefully this fixes your build. Can you confirm?
>

Works like a charm - ran multiple builds, all OK, including "fixing" my
earlier bad builds, and various clean new builds ran right to completion.
Excellent.
